gems/smithy/lib/smithy/welds/synthetic_input_output.rb

```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+module Smithy
+  module Welds
+    # Creates synthetic input and output shapes for operations that do not have them.
+    class SyntheticInputOutput < Weld
+      def pre_process(model)
+        say_status :insert, 'Creating synthetic input and output shapes', @plan.quiet
+        create_synthetic_input_output_shapes(model)
+      end
+
+      private
+
+      def create_synthetic_input_output_shapes(model)
+        operations = Model::ServiceIndex.new(model).operations_for(@plan.service)
+        operations.each do |operation_id, operation|
+          create_synthetic_input_shape(model, operation_id, operation) if operation['input']
+          create_synthetic_output_shape(model, operation_id, operation) if operation['output']
+        end
+      end
+
+      def create_synthetic_input_shape(model, operation_id, operation)
+        input_target = operation['input']['target']
+        target = Model.shape(model, input_target)
+        return if target.fetch('traits', {}).include?('smithy.api#input') || input_target == 'smithy.api#Unit'
+
+        input_shape = {
+          'type' => 'structure',
+          'traits' => target.fetch('traits', {}).merge({ 'smithy.api#input' => {} }),
+          'members' => target.fetch('members', {}).merge({})
```

Comment:
Do we need to do a deep copy of members and traits so that another weld that modifies the synthetic shape don't modify the original.  I know the merge will do a shallow copy and I can't remember the details of how these are structured so this may not apply.

Reply:
Yes. This was on my radar. We need a deep copy.